### ISSUE TITLE
Error when compiling my custom fork of leveldown (to use zlib instead of snappy)

### ISSUE THREAD

**Issue by clarkx86** [Level/leveldown#629]
Hello there :-),

I'm trying to compile my [custom fork](https://github.com/clarkx86/leveldown/commits/) of `leveldown`, which I tried to configure to use [Mojang's fork](https://github.com/Mojang/leveldb-mcpe) of LevelDB, which uses zlib compression instead of snappy, in order to replace the native LevelDB dependency.

Unfortunately, I don't know much C++ and got referenced to [this thread](https://github.com/Level/level/issues/119#issuecomment-459327002) by @vweevers. I did every change to the `binding.gyp`, `binding.cc`, `leveldb.gyp` mentioned in the thread and replaced the LevelDB dependency in `./deps/leveldb/leveldb-1.20/` with Mojang's version.

I ran into some more troubles however, and get thrown the following error when trying to install (`npm install clarkx86/leveldown`):

![grafik](https://user-images.githubusercontent.com/46745823/57198514-f858c480-6f73-11e9-88a6-d0aeb957585e.png)
![grafik](https://user-images.githubusercontent.com/46745823/57198621-2ab6f180-6f75-11e9-8c6d-a8559a0e1cf6.png)

This is an excerpt of [the full crashlog](https://gist.github.com/clarkx86/4b7e34d0cba7b8aa7319954de08a3d2f).
I tried to compile on Ubuntu. On Windows it seems to spit out [some different errors](https://gist.github.com/clarkx86/6edc63eba03f0556d862196b79f60e96).

You can view my fork of `leveldown` and the changes I did [here](https://github.com/clarkx86/leveldown/commits/master).

In case of the compiling on Ubuntu, it seems to adress `slice.h`. I compared the `slice.h` of the original dependency with the one of Mojang's fork, but the only difference is:

![grafik](https://user-images.githubusercontent.com/46745823/57198961-91d6a500-6f79-11e9-8e08-d0393c4a10fa.png)

Which I resolved by adding
```
"defines": [
      "SNAPPY=0",
      "DLLX="
]
```
in `leveldb.gyp`.

To be fair, when compiling the native `leveldown` repository, everything works fine :), so it must be an issue with my fork.

I'm sorry if this is *a bit* off-topic, but I really hope that someone would have an idea to get to the root of this problem.

Thank you and have a wonderful day,
clarkx86 :).

**Comment by ralphtheninja** [comment]
> In case of the compiling on Ubuntu, it seems to adress slice.h. I compared the slice.h of the original dependency with the one of Mojang's fork, but the only difference is:

It might be worth it to look at the complete diff than just checking `slice.h`. There might be more to it than "just" setting `DLLX` (which looks correct btw).

**Comment by vweevers** [comment]
Maybe @peakji has an idea. Or the folks in the Mojang repo - though it'll also be off-topic there 😄

**Comment by clarkx86** [comment]
Thanks for your reply @ralphtheninja & @vweevers :)!
I did do a complete difference-check using WinMerge. The majority of the source code is actually untouched with slight changes like these (for example):
![grafik](https://user-images.githubusercontent.com/46745823/57199938-580b9b80-6f85-11e9-8d83-fb7f07acbad2.png)
or this
![grafik](https://user-images.githubusercontent.com/46745823/57199971-ac168000-6f85-11e9-9670-627270d4c603.png)
or added `DLLX` definitions before most classes (there are of course also a few more serious changes, but I think just taking screenshots of the compared code and dumping them here doesn't help anyone at all, haha :D).

The `uint32_t` definitions shouldn't be a problem in terms of compiling (?), please correct me if I'm wrong (please take in mind that I really only know *veeery* few C++).

I can also try opening an issue thread over at Mojang's repo and see if the folks can come up with a hint in the right direction, just thought posting here might be a good place to start :).

Thank you very much in advance so far guys!
- clarkx86

**Comment by ralphtheninja** [comment]
@clarkx86 Aah, I see now. Will submit a fix to your fork.

**Comment by ralphtheninja** [comment]
See https://github.com/clarkx86/leveldown/pull/1

**Comment by clarkx86** [comment]
**Wow**! Thank you so much @ralphtheninja! You're a hero!
Merged your PR and it does compile now successfully!
Simple, yet amazing fix :D. Sometimes it's the small things ^^.

The compile log notifies me about some warnings:
![grafik](https://user-images.githubusercontent.com/46745823/57200909-973fe980-6f91-11e9-82b3-e4e029ea524e.png)
![grafik](https://user-images.githubusercontent.com/46745823/57200912-a32bab80-6f91-11e9-8f0c-a925413ed96b.png)
Would this be a problem at first glance?

Also, I don't want to ask for too much from you guys, but I just can't seem to compile for Windows **on** Windows for some reason ([crash log](https://gist.github.com/clarkx86/6edc63eba03f0556d862196b79f60e96)). That error log is completely out of my scope (and I assume most likely also of yours), would it be possible to cross-compile on Ubuntu for Windows (and just out of pure curiosity, macOS?). And if so, what would be the procedure to do so? I saw some `cross-compile` script (for Travis CI?) in the `scripts` directory.

Again, thank you **very** much! You made my day!

**Comment by vweevers** [comment]
> Would this be a problem at first glance?

It's fine. We have the same warning with LevelDB (https://github.com/Level/leveldown/issues/568)

> Would it be possible to cross-compile on Ubuntu for Windows

In theory yes! Though I'm not sure it would solve this particular problem. That said, we use docker images based on [`dockcross`](https://github.com/dockcross/dockcross) to cross-compile and that projects also offers docker images to cross-compile for Windows. Our [images](https://github.com/prebuild/docker-images) are a thin layer on top of `dockcross` images to 1) install node and 2) set some environment variables.

> for Travis CI?

It works both locally (requires `docker` to be installed) and in Travis. We use Travis to automatically make prebuilt binaries whenever we release a new version.

**Comment by clarkx86** [comment]
> Though I'm not sure it would solve this particular problem.

I guess not actually, haha! Thank you anyways, you guys helped me a ton! You rock & keep it up :) 👍!

Regarding the Windows compilation issue, if anyone stumbles upon this thread and feels like giving it a go in their spare time, I would really appreciate hearing about some progress! I got the feeling that it could indeed be achievable.

Otherwise, thanks once again to @ralphtheninja and @vweevers for your help! Have an awesome start into the week ;)!

-clarkx86

**Comment by clarkx86** [comment]
Sorry, I have to dig out this thread again to ask one more question someone might have an answer to. I successfully compiled the bindings for Ubuntu, but I get this error message when **requiring** my `leveldown` fork. Of course, this is absolutely exclusive to my fork, not the main leveldown fork :)!
```
Error: /mnt/c/Users/clark/Cloud/Google Drive/Dokumente/GitHub/Projects/node.js/_playground/playground_zlib/node_modules/leveldown/build/Release/leveldown.node: undefined symbol: _ZNK7leveldb18ZlibCompressorBase12compressImplEPKcmRNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE
    at Object.Module._extensions..node (internal/modules/cjs/loader.js:730:18)
    at Module.load (internal/modules/cjs/loader.js:600:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:539:12)
    at Function.Module._load (internal/modules/cjs/loader.js:531:3)
    at Module.require (internal/modules/cjs/loader.js:637:17)
    at require (internal/modules/cjs/helpers.js:22:18)
    at load (/mnt/c/Users/clark/Cloud/Google Drive/Dokumente/GitHub/Projects/node.js/_playground/playground_zlib/node_modules/leveldown/node_modules/node-gyp-build/index.js:20:10)
    at Object.<anonymous> (/mnt/c/Users/clark/Cloud/Google Drive/Dokumente/GitHub/Projects/node.js/_playground/playground_zlib/node_modules/leveldown/binding.js:1:105)
    at Module._compile (internal/modules/cjs/loader.js:701:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:712:10)
```

Basically, yeah: undefined symbol: `undefined symbol: _ZNK7leveldb18ZlibCompressorBase12compressImplEPKcmRNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE`

I think this could be a problem of how the options for compression are defined in the modified `bindings.cc` [here](https://github.com/clarkx86/leveldown/commit/cf6b04e7c9eef1aa7a2f8ec392f03fa449def192#diff-0003efb78f2f135c88ca5431b8353b2c)? I did specify the `#include <leveldb/zlib_compressor.h>` headers.

**Comment by vweevers** [comment]
Closing, because it's not about `leveldown` and although we're willing to help, further help would probably require a deep dive into the fork.

**Comment by clarkx86** [comment]
First of all, sorry to post in this closed issue once again.
However, I feel like I nearly got it working now and I've run into an issue you might have an idea for. Compiling and requiring the module works without any problems, but when opening a database within node.js I get this error before the process aborts:
```
node: ../deps/leveldb/leveldb/db/version_set.cc:798: leveldb::VersionSet::~VersionSet(): Assertion `dummy_versions_.next_ == &dummy_versions_' failed.
```

I saw that this was a previously known issue and I've read about it here https://github.com/Level/leveldown/issues/194 for example, but I just can't seem to wrap my head around this...

Again, please ignore this post if it's too off-topic. Just asking in the hope anyone has a slight clue what this could mean. :)

Thank you very much and have an awesome day,
Clark :)!

**Comment by ralphtheninja** [comment]
@clarkx86 Check out this patch https://github.com/Level/leveldown/pull/368/files#diff-155c57a0315eaa7e661e1b190e7ea520R155 .. does your fork have the same fix?

**Comment by clarkx86** [comment]
Hey @ralphtheninja, thanks for your quick answer! Really appreciate it :).

My fork is on the same _level_ (pun intended :D) as the current master, so I checked if the patch in the `src/iterator.cc` you've mentioned above matched with what is in the corresponding lines of the [`binding.cc`](https://github.com/clarkx86/leveldown-zlib/blob/036ff281e98a41df753534fbb28e05f9ea17d193/binding.cc#L686-L695) in my fork (I noticed things changed a lot since then of course) and it looks pretty similar.
